staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/strategy.go

```diff
@@ -59,6 +59,12 @@ func (strategy) PrepareForCreate(ctx context.Context, obj runtime.Object) {
 	if !utilfeature.DefaultFeatureGate.Enabled(apiextensionsfeatures.CustomResourceValidation) {
 		crd.Spec.Validation = nil
 	}
+	if !utilfeature.DefaultFeatureGate.Enabled(apiextensionsfeatures.CustomResourceValidation) ||
+		!utilfeature.DefaultFeatureGate.Enabled(apiextensionsfeatures.CustomResourceWebhookConversion) {
```

Comment:
If we drop the dependency on `apiextensionsfeatures.CustomResourceWebhookConversion`, leaving v.Schema non-nil will fail validation https://github.com/kubernetes/kubernetes/pull/67553/commits/63040fd4f87cd8ab8c75e6b3b0eca702205c89bf#diff-54fdd67617b0f2ef1b5a09148f77303eR107. 

I feel we need to drop the validation dependency on `WebhookConversion` as well. I can imagine leaving v.Schema non-nil would not harm in a single version. The problem is if a user rolls forward from version X of apiserver that disables the feature to version Y that enables it, per-version schema will be honored: 
 - If we prune per-version schema in version X, CR creation in version Y will have no validation schema. 
 - If we do not prune per-version schema in version X, in version Y the per-version schema will be either user-specified value, or defaulted value (which is same as the top level validation schema)

Reply:
This is a little tricky and I think @erictune will have opinion here too.

Here we do prune per version schema for create and update. So if we enable/disable feature gate (in a roll-back/forward) we do not lose these schemas if we do not update CRDs.

However on create an update, if user send an invalid Schemas per version to an API server that has the featureGate == false, and if we do not prune that, we will have a problem when we enable the feature and fixing it would be a headache for admins. Of course we can also validate those schemas regardless of feature gate value but I do not recommend that.

I think we should prune per version schema if the feature gate is disabled, after all this would be the behaviour if the API server does not have Schema per version at all (1.11).

Reply:
I agree it is the behavior if the API server does not have Schema per version at all. And I feel doing the [defaulting  even if `CustomResourceWebhookConversion` is disabled](https://github.com/kubernetes/kubernetes/pull/67553/commits/4190185592454141051de291f16e80ab46df2807) may be doing too much on users' behalf. 

But I'm still concerned about the rolling forward story. If we prune per version fields (schema, subresources, columns) when `CustomResourceWebhookConversion` is disabled, when we enable the feature the per version subresources (which are nil) will be honored, so the versions other than the top level version will stop serving subresources. Is this intended behavior?

Reply:
Defaulting is happening on the read path, right? so making them nil should be fine. When the featuregate is enabled, it should default all versions (except first one) with top level one. This defaulting should also happen only if the feature gate is on.

Reply:
Ah, sorry I didn't realize defaulting also kicks in when we update apiserver to enable the feature, as it happens on decoding storage data. The rolling forward story sounds good to me then. I updated the code. The defaulting, validation, strategy(pruning) are all feature gated now